### PR TITLE
feature: score switching to general score types - add option to perform protein inference on non-main scores

### PR DESCRIPTION
**Pull request by timosachsenberg**
**Motivation**
ID results often contain either:
- A search engine-specific score as the main score (e.g., "X!Tandem" scores), or
- A broader score category, such as posterior error probabilities (PEP) or q-values.

In practice, many algorithms require specific types of scores, such as probabilities, for compatibility (e.g., PEP or PP). To address this, the IDScoreSwitcherAlgorithm has been extended to allow dynamic switching of the main score to a requested score type. This ensures compatibility across algorithms while providing the ability to revert back to the original main score when the process is complete.

**Algorithm:**
New Parameter for BasicProteinInference: score_type. 
Previously, the BPI always relied on the main PSM score for inference. This update introduces a parameter that enables the selection of a score from the meta values annotated to the PSMs (PeptideHit), rather than being restricted to the main score.
If the requested score is already the main score, no action is needed.
Otherwise, the IDScoreSwitcherAlgorithm is used to find and switch the score of the requested type/category to the main score.
After processing, the main score is reverted to its original value.
Enhancements to IDScoreSwitcherAlgorithm: To generalize and simplify this functionality for wider use, the logic has been moved into the IDScoreSwitcherAlgorithm. The algorithm now supports two new static methods:

switchToScoreType: Accepts a string representing the score type (e.g., "RAW", "PEP", "q-value") and switches the main score accordingly.
switchBackScoreType: Reverts the main score to its previous state using the result object from the initial switch.
Benefits

**ProteinInference Tool:**
By adding the score switching logic to the BasicProteinInference algorithm the TOPP tool ProteinInference gained a new parameter.
```
ProteinInference \
    -in ID_mapper_merge.consensusXML \
    -threads 16 \
    -picked_fdr true \
    -picked_decoy_string DECOY_ \
    -protein_fdr true \
    -Algorithm:use_shared_peptides true \
    -Algorithm:annotate_indistinguishable_groups true \
    -Algorithm:score_aggregation_method best \
    -Algorithm:min_peptides_per_protein 1 \
    -out ID_mapper_merge_epi_new.consensusXML \
    -Algorithm:score_type "PEP"     <-------------------------------- this is new
```

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: src/tests/class_tests/openms/data/newMergerTest_out.idXML
Comment (timosachsenberg) [inline src/tests/class_tests/openms/data/newMergerTest_out.idXML:1]: I removed the XTandem scores from proteins and put them on the PSMs

File: src/tests/class_tests/openms/data/PeptideProteinResolution_out2_ibg.idXML [hunk 1]
```diff
@@ -36,53 +36,53 @@
...
 				<UserParam type="float" name="E-Value" value="20.0"/>
 				<UserParam type="float" name="nextscore" value="8.6"/>
 			</PeptideHit>
 		</PeptideIdentification>
 		<PeptideIdentification score_type="Posterior Probability" higher_score_better="true" significance_threshold="0.0" MZ="784.913208007812955" RT="2399.082299999980023" >
 			<PeptideHit score="0.1" sequence="DDD" charge="2" aa_before="R X" aa_after="C X" protein_refs="PH_3 PH_4" >
 				<UserParam type="string" name="target_decoy" value="target"/>
-				<UserParam type="float" name="XTandem_score" value="8.9"/>
 				<UserParam type="string" name="protein_references" value="non-unique"/>
+				<UserParam type="float" name="XTandem_score" value="8.9"/>
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/data/PeptideProteinResolution_out2_ibg.idXML:58]: not sure why the order changed here

**Comment by jpfeuffer** [comment]
I usually prefer allowing choosing score TYPES. No one is able to remember the horrible CV terms that we often store.

**Comment by timosachsenberg** [comment]
how would that work here? I thought support should also be for custom raw scores.
But ok if I add enum strings I could search for meta values that match the type I guess. 
so you mean basically only using:
switchToGeneralScoreType?
Honestly it is quite a mess how "types" and "names" are used so I probably missed that possibility.

File: src/openms/include/OpenMS/ANALYSIS/ID/IDScoreSwitcherAlgorithm.h [hunk 2]
```diff
@@ -52,8 +81,19 @@ namespace OpenMS
...
+      This static method attempts to map a given string, representing a score type, to the corresponding
+      ScoreType enum value. It handles various common representations of score types, including those
+      with or without the "_score" suffix, and ignores case and special characters like '-', '_', and ' '.
+
+      @param score_type The string representation of the score type.
+      @return The corresponding ScoreType enum value.
+      @throws Exception::MissingInformation If the provided score_type string does not match any known
+                                            score type.
+    */
+    static ScoreType toScoreTypeEnum(String score_type)
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/IDScoreSwitcherAlgorithm.h:96]: just wanted to have something different from getScoreType ;)

File: src/openms/source/ANALYSIS/ID/BasicProteinInferenceAlgorithm.cpp [hunk 3]
```diff
@@ -59,6 +58,11 @@ namespace OpenMS
     defaults_.setValue("greedy_group_resolution", "false", "If this is true, shared peptides will be associated to best proteins only (i.e. become potentially quantifiable razor peptides).");
     defaults_.setValidStrings("greedy_group_resolution", {"true","false"});
 
+    defaults_.setValue("score_type", "", "PSM score type to use for inference. (default: empty = main score)");
+    defaults_.setValidStrings("score_type", {"", "PEP", "q-value", "RAW"});
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/BasicProteinInferenceAlgorithm.cpp:62]: now as yasset suggested, "" just takes the main score as before

File: src/tests/class_tests/openms/source/BasicProteinInferenceAlgorithm_test.cpp [hunk 4]
```diff
@@ -140,4 +140,43 @@ START_TEST(BasicProteinInferenceAlgorithm, "$Id$")
...
+      p.setValue("annotate_indistinguishable_groups", "true");
+      p.setValue("greedy_group_resolution", "true");
+      p.setValue("score_type", "RAW");  // should use the XTandem score meta value
+      bpia.setParameters(p);
+
+      TEST_EQUAL(peps[0].getScoreType(), "Posterior Error Probability"); // check if main score is PEP
+      bpia.run(peps, prots);
+      TEST_EQUAL(peps[0].getScoreType(), "Posterior Error Probability"); // check if main score has been reset again to PEP
+
+      TEST_EQUAL(prots[0].getHits().size(), 5)
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/BasicProteinInferenceAlgorithm_test.cpp]: this is one more than before???
is this expected when switching to the RAW score?

File: src/tests/class_tests/openms/data/newMergerTest_out.idXML [hunk 5]
```diff
@@ -11,19 +11,15 @@
 		<ProteinIdentification score_type="" higher_score_better="false" significance_threshold="0" >
 			<ProteinHit id="PH_0" accession="A" score="0.258519381284714" sequence="" >
 				<UserParam type="string" name="target_decoy" value="target"/>
-				<UserParam type="float" name="XTandem_score" value="0"/>
```
Comment (jpfeuffer) [inline src/tests/class_tests/openms/data/newMergerTest_out.idXML:14]: I would leave this in, since unfortunately this is how it would come out of our pipelines.

File: src/tests/class_tests/openms/data/newMergerTest_out.idXML [hunk 6]
```diff
@@ -37,42 +33,50 @@
...
 				<UserParam type="string" name="protein_references" value="unique"/>
+				<UserParam type="float" name="XTandem_score" value="10"/>
 			</PeptideHit>
 			<UserParam type="int" name="id_merge_index" value="1"/>
 		</PeptideIdentification>
 		<PeptideIdentification score_type="Posterior Error Probability" higher_score_better="false" significance_threshold="31.8621" MZ="0" RT="0" >
 			<PeptideHit score="0.1" sequence="FIANGE" charge="2" aa_before="D" aa_after="R" start="1" end="6" protein_refs="PH_5" >
 				<UserParam type="string" name="target_decoy" value="target"/>
 				<UserParam type="string" name="protein_references" value="unique"/>
+				<UserParam type="float" name="XTandem_score" value="20"/>
```
Comment (jpfeuffer) [inline src/tests/class_tests/openms/data/newMergerTest_out.idXML]: You could have used similar/same scores for testing, to see if you get the same output.
I didn't go through the graph yet to see if the new number makes sense.

File: src/openms/source/ANALYSIS/ID/BasicProteinInferenceAlgorithm.cpp [hunk 7]
```diff
@@ -59,6 +58,11 @@ namespace OpenMS
...
+
+    defaults_.setValidStrings("greedy_group_resolution", {"true","false"});
```
Comment (jpfeuffer) [inline src/openms/source/ANALYSIS/ID/BasicProteinInferenceAlgorithm.cpp]: Looks like a duplicated line on my diff.

File: src/tests/class_tests/openms/data/newMergerTest_out.idXML (hunk 5 above)
Comment (timosachsenberg) [reply src/tests/class_tests/openms/data/newMergerTest_out.idXML:14]: ah why again at the ProteinHit?

File: src/tests/class_tests/openms/data/newMergerTest_out.idXML (hunk 6 above)
Comment (timosachsenberg) [reply src/tests/class_tests/openms/data/newMergerTest_out.idXML]: that's a great point. I will change it to have similar but different scores

File: src/openms/include/OpenMS/ANALYSIS/ID/IDScoreSwitcherAlgorithm.h [hunk 8]
```diff
@@ -552,6 +552,145 @@ namespace OpenMS
...
+    IDScoreSwitcherAlgorithm::ScoreType original_score_type = IDScoreSwitcherAlgorithm::ScoreType::RAW;
+    // the score name, orientation and type used after the switch
+    bool requested_score_higher_better = original_score_higher_better;    
+    IDScoreSwitcherAlgorithm::ScoreType requested_score_type = original_score_type;      
+    String requested_score_name; // the search engine score name (e.g. "X!Tandem_score" or score category (e.g. "PEP")
+    // wheter the main score was switched
+    bool score_switched = false;
+  };
+
+  static IDSwitchResult switchToScoreType(ConsensusMap& cmap, String requested_score_type_as_string, bool include_unassigned = true)
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/IDScoreSwitcherAlgorithm.h:592]: moved to IDScoreSwitcher for general use.
now has:
switchToScoreType
switchBackScoreType

to set and reset a score type

File: src/tests/class_tests/openms/data/newMergerTest_out.idXML [hunk 9]
```diff
@@ -37,42 +33,50 @@
 			<PeptideHit score="0.4" sequence="STFDFYQRRLVTLAESPRAPSP" charge="2" aa_before="K [" aa_after="G ]" start="227 4" end="248 25" protein_refs="PH_0 PH_1" >
 				<UserParam type="string" name="target_decoy" value="decoy"/>
 				<UserParam type="string" name="protein_references" value="non-unique"/>
+				<UserParam type="float" name="XTandem_score" value="2.5"/>
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/data/newMergerTest_out.idXML:36]: I set XTandem_score = 1.0/PEP of the original test

File: src/tests/class_tests/openms/data/newMergerTest_out.idXML (hunk 5 above)
Comment (jpfeuffer) [reply src/tests/class_tests/openms/data/newMergerTest_out.idXML:14]: It comes from score switching and new score/fdr calculations. We always store the old score, even if it is meaningless.

**Comment by timosachsenberg** [comment]
These are the PEP test results:
```
      TEST_EQUAL(prots[0].getHits().size(), 4)
      TEST_EQUAL(prots[0].getHits()[0].getScore(), 0.6)
      TEST_EQUAL(prots[0].getHits()[1].getScore(), 0.6)
      TEST_EQUAL(prots[0].getHits()[2].getScore(), 0.8)
      TEST_EQUAL(prots[0].getHits()[3].getScore(), 0.9)

      TEST_EQUAL(prots[0].getHits()[0].getMetaValue("nr_found_peptides"), 1)
      TEST_EQUAL(prots[0].getHits()[1].getMetaValue("nr_found_peptides"), 1)
      TEST_EQUAL(prots[0].getHits()[2].getMetaValue("nr_found_peptides"), 2)
      TEST_EQUAL(prots[0].getHits()[3].getMetaValue("nr_found_peptides"), 1)

      TEST_EQUAL(prots[0].getIndistinguishableProteins().size(), 3);
      TEST_EQUAL(prots[0].getIndistinguishableProteins()[0].probability, 0.9);
      TEST_EQUAL(prots[0].getIndistinguishableProteins()[1].probability, 0.8);
      TEST_EQUAL(prots[0].getIndistinguishableProteins()[2].probability, 0.6);
```

These are the XTandem score test results:
```
      TEST_EQUAL(prots[0].getHits().size(), 4)
      TEST_EQUAL(prots[0].getHits().at(0).getScore(), 2.5)
      TEST_EQUAL(prots[0].getHits().at(1).getScore(), 2.5)
      TEST_EQUAL(prots[0].getHits().at(2).getScore(), 2.5) 
      TEST_EQUAL(prots[0].getHits().at(3).getScore(), 10.0)

      TEST_EQUAL(prots[0].getHits().at(0).getMetaValue("nr_found_peptides"), 1)
      TEST_EQUAL(prots[0].getHits().at(1).getMetaValue("nr_found_peptides"), 1)
      TEST_EQUAL(prots[0].getHits().at(2).getMetaValue("nr_found_peptides"), 2)
      TEST_EQUAL(prots[0].getHits().at(3).getMetaValue("nr_found_peptides"), 1)    

      TEST_EQUAL(prots[0].getIndistinguishableProteins().size(), 3);
      TEST_EQUAL(prots[0].getIndistinguishableProteins().at(0).probability, 10);
      TEST_EQUAL(prots[0].getIndistinguishableProteins().at(1).probability, 2.5);
      TEST_EQUAL(prots[0].getIndistinguishableProteins().at(2).probability, 2.5);  
```

I used XTandem score = 1/PEP to introduce a meta value.

Sizes of groups etc. are correct now but I think there is something of with the "probabilities" and scores.

File: src/tests/class_tests/openms/data/newMergerTest_out.idXML (hunk 5 above)
Comment (jpfeuffer) [reply src/tests/class_tests/openms/data/newMergerTest_out.idXML:14]: But yeah maybe you can leave it out. It is just a corner case to test maybe. Usually proteins are not touched before inference.

**Comment by jpfeuffer** [comment]
Yes, something is wrong. Probably score direction.

**Comment by timosachsenberg** [comment]
I added another test.
@jpfeuffer if you find time can you take a look at the inference results in the test?

**Comment by jpfeuffer** [comment]
In this case it should be higher = better, no?

**Comment by timosachsenberg** [comment]
yes. for raw scores it should always: higher = better

**Comment by jpfeuffer** [comment]
But then something is wrong. It obviously prefers 2.5 over 5 for PH3

**Comment by timosachsenberg** [comment]
Might have found a bug...
switching from a lower is better to a higher is better raw score did not update the higher is better...

File: src/openms/include/OpenMS/ANALYSIS/ID/IDScoreSwitcherAlgorithm.h [hunk 10]
```diff
@@ -260,9 +260,9 @@ namespace OpenMS
       }
       new_score_ = t;
 
-      if (type != ScoreType::RAW && higher_better_ != type_to_better_[type])
+      if (higher_better_ != type_to_better_[type])
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/IDScoreSwitcherAlgorithm.h:263]: we currently enforce that all raw scores are higher better so I fix this here as well

**Comment by timosachsenberg** [comment]
Looks correct now:
```
      TEST_EQUAL(prots[0].getHits().size(), 4)
      TEST_EQUAL(prots[0].getHits().at(0).getScore(), 2.5)
      TEST_EQUAL(prots[0].getHits().at(1).getScore(), 2.5)
      TEST_EQUAL(prots[0].getHits().at(2).getScore(), 5.0) 
      TEST_EQUAL(prots[0].getHits().at(3).getScore(), 10.0)

      TEST_EQUAL(prots[0].getHits().at(0).getMetaValue("nr_found_peptides"), 1)
      TEST_EQUAL(prots[0].getHits().at(1).getMetaValue("nr_found_peptides"), 1)
      TEST_EQUAL(prots[0].getHits().at(2).getMetaValue("nr_found_peptides"), 2)
      TEST_EQUAL(prots[0].getHits().at(3).getMetaValue("nr_found_peptides"), 1)    

      TEST_EQUAL(prots[0].getIndistinguishableProteins().size(), 3);
      TEST_EQUAL(prots[0].getIndistinguishableProteins().at(0).probability, 10);
      TEST_EQUAL(prots[0].getIndistinguishableProteins().at(1).probability, 5.0);
      TEST_EQUAL(prots[0].getIndistinguishableProteins().at(2).probability, 2.5);  
```

**Review by jpfeuffer — APPROVED** [review]
Yes tests look good now. Also your large scale test in our chat sounds correct. Code overview also looks fine. Couldn't check exact details, though, yet.

**Comment by jpfeuffer** [comment]
Long term we could think about other defaults.  Yes, in theory most people will want to use PEP or PP. But it is good to have it as a general aggregation engine.
